### ISSUE TITLE
How to use HtmlForm to submit and how to use jQuery

### ISSUE THREAD

**Issue by zhenxi1023** [HtmlUnit/htmlunit#307]
I want to use HtmlUnit to help me to submit a search form in best buy. And I could use these jQuery code in Chrome:
$("#gh-search-input").val("Apple"); $("[name=frmSearch]").submit();
And the Chrome could jump to a new page: https://www.bestbuy.com/site/brands/apple/pcmcat128500050005.c?id=pcmcat128500050005.
How could I do the same thing by HtmlUnit?
And, Best Buy page use jQuery, may I use jQuery to search dom and use event callback? How could use in the HtmlUnit?
Thanks

**Comment by rbri** [comment]
In general you can do the same with HtmlUnit - some hints here

* get you page
* getElementById("gh-search-input")
* type("Apple)
* locate the form (have a look at https://htmlunit.sourceforge.io/gettingStarted.html there is a section 'Finding a specific element'
* submit the form

maybe you have to wait a bit for all the ajax stuff to finish

**Comment by zhenxi1023** [comment]
Well, thanks for your reply. But the problem still have.
About the submit form, I use these code to submit the form, but I can't get the new page. jQuery code $("[name=frmSearch]").submit(); about $("[name=frmSearch]"), this is a form element, not a button. And after my experience, I think the button in the form without any click event bundled, they just using form action. Thus, how could I use by HtmlUnit to submit the form? 
And I guess I could use HtmlForm.submit method, but how could I write the code to get the new page?
Here I show my code about them.

```
import com.gargoylesoftware.htmlunit.BrowserVersion;
import com.gargoylesoftware.htmlunit.WebClient;
import com.gargoylesoftware.htmlunit.html.*;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import javax.net.ssl.SSLHandshakeException;
import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Test {

   public static void main(String[] args) throws IOException {

      HtmlPage page = getPage("https://www.bestbuy.com/");
      HtmlInput input = (HtmlInput) page.getElementById("gh-search-input");
      input.type("Apple");
      HtmlForm htmlForm = page.getFormByName("frmSearch");
      for (DomElement childElement : htmlForm.getChildElements()) {
         if (childElement.getNodeName().equals("button") && childElement.getAttribute("title").equals("submit search")) {
            HtmlPage newPage = childElement.click();
            System.out.println(newPage.getUrl()); // finally, I get the url is still https://www.bestbuy.com, and when use jQuery in the Chrome, the page url will become: https://www.bestbuy.com/site/brands/apple/pcmcat128500050005.c?id=pcmcat128500050005 . So I think submit search failed.
         }
      }
//      SubmittableElement element = (SubmittableElement) page.getByXPath("button[@class='header-search-button']");
//      htmlForm.submit(element);

   }


   public static HtmlPage getPage(String url) throws IOException {
      WebClient webClient = new WebClient(BrowserVersion.CHROME);
      webClient.getOptions().setThrowExceptionOnScriptError(false);
      webClient.getOptions().setThrowExceptionOnScriptError(false);
      webClient.getOptions().setActiveXNative(false);
      webClient.getOptions().setCssEnabled(false);
      webClient.getOptions().setJavaScriptEnabled(false);
      webClient.waitForBackgroundJavaScript(1000);
      try {
         HtmlPage page = webClient.getPage(url);
         return page;
      } finally {
         webClient.close();
      }
   }

}
```


And here is the dom tree of Best Buy:



Thanks

**Comment by zhenxi1023** [comment]
```
<form method='GET' name='frmSearch' action='https://www.bestbuy.com/site/searchpage.jsp'>
<input id='gh-search-input'/>
<button><svg>...</svg></button>  // clear button
<button type='submit' class='header-search-button' title='submit search'><svg>...</svg></button>  // search button
</form>
```

**Comment by rbri** [comment]
Ok, there are some problems with you code

* the WebClient is the same as your browser - closing the client after getting the page is not a good idea - you have to close the client if your are done, not earlier
* webClient.waitForBackgroundJavaScript(1000); is not a option, you have to do this call after e.g. click()
* HtmlPage newPage = childElement.click(); is only the sync part of the click(). I fear in your case some async js might run and replaces the page by another one. To be on the save side try something like this


    HtmlPage page = childElement.click();
    webClient.waitForBackgroundJavaScript(1000);
    page = (HtmlPage) page.getEnclosingWindow().getEnclosedPage();

**Comment by zhenxi1023** [comment]
Thank you for your answer. It's going to work. But about these code: webClient.waitForBackgroundJavaScript(1000); and execute JavaScript, I have more questions.
	As we known, most JavaScript is async. So when we write code always use a callback function. May I use the same way, callback function, to execute 
page=(HtmlPage) page.getEnclosingWindow().getEnclosedPage(); 
Cause I setting wait time too small to couldn't finish the task, or else.
   I found a method that is 
page.executeJavaScript(String code);
I think this is really nice. But I have to set options, 
WebClinet.getOptions().setJavaScriptEnabled(true)
Because the page I want to load have script too. This setting will waste more time, is there have a way to choose which script could use?
Thank you very much!

**Comment by rbri** [comment]
Not sure i got your point - will try to make it more clear

* most of the action methods (e.g. click) are returning the sync result. This means, if you have a button having an onClickHandler then the HtmlUnit click() method returns after the handler is executed. All further steps are seeing the page at this time (ok, the js engine still runs in parallel and might update your page while your are processing your code). If the on-handler forces some async js (setTimer, Promise, ...) then you have to wait. All updates of the Dom done from js during this wait are visible in the page object you have in your hand.
* some javascript forces the page to reload inside the window by e.g. setting the location. This kind of reload are building a new page object that is placed in the window. In this case your code still points to the old page object. In this cases you have to refresh your page reference by asking the window for the (new) page.

Hope that helps....

**Comment by zhenxi1023** [comment]
Well, I mean, the HtmlPage which I requested, has some scripts. I could use: `webClient.getOptions().setJavaScriptEnabled(false); `
to disabled them to load. This will save more time. But I saw, a method is 
`page.executeJavaScript("document.getElementById('gh-search-input')");`
If I want to use this code, I must set, 
`webClient.getOptions().setJavaScriptEnabled(true); `
Thus, I want to know, is there a way, to avoid loading the requested page's scripts, but could execute the js code by this way
`page.executeJavaScript("document.getElementById('gh-search-input')");`
Thanks
